crates/re_sdk/src/recording_stream.rs

```diff
@@ -614,7 +616,59 @@ impl RecordingStreamBuilder {
 /// Shutting down cannot ever block.
 #[derive(Clone)]
 pub struct RecordingStream {
-    inner: Arc<Option<RecordingStreamInner>>,
+    inner: Either<Arc<Option<RecordingStreamInner>>, Weak<Option<RecordingStreamInner>>>,
+}
+
+impl RecordingStream {
+    /// Passes a reference to the [`RecordingStreamInner`], if it exists.
+    ///
+    /// This works whether the underlying stream is strong or weak.
+    #[inline]
+    fn with<F: FnOnce(&RecordingStreamInner) -> R, R>(&self, f: F) -> Option<R> {
+        use std::ops::Deref as _;
+        match &self.inner {
+            Either::Left(strong) => strong.deref().as_ref().map(f),
+            Either::Right(weak) => weak
+                .upgrade()
+                .and_then(|strong| strong.deref().as_ref().map(f)),
+        }
+    }
+
+    /// Clones the [`RecordingStream`] without incrementing the refcount.
+    ///
+    /// Useful e.g. if you want to make sure that a detached thread won't prevent the [`RecordingStream`]
+    /// from flushing during shutdown.
+    //
+    // TODO(#5335): shutdown flushing behavior is too brittle.
+    #[inline]
+    pub fn clone_weak(&self) -> Self {
+        Self {
+            inner: match &self.inner {
+                Either::Left(strong) => Either::Right(Arc::downgrade(strong)),
+                Either::Right(weak) => Either::Right(Weak::clone(weak)),
+            },
+        }
+    }
+}
+
+// TODO(#5335): shutdown flushing behavior is too brittle.
+impl Drop for RecordingStream {
+    #[inline]
+    fn drop(&mut self) {
+        // If this holds the last strong handle to the recording, make sure that all pending
+        // `DataLoader` threads started from the SDK actually run to completion.
+        //
+        // NOTE: It's very important to do so from the `Drop` implementation of `RecordingStream`
+        // itself, because the dataloader threads -- by definition -- will have to send data into
+        // this very recording, therefore we must make sure that at least one strong handle still lives
+        // on until they are all finished.
```

Comment:
would have made it easier for me to understand if you noted here that the dataloader threads have weak handles